exporter/prometheusexporter/exporter.go

```diff
@@ -0,0 +1,350 @@
+// Copyright 2019, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheusexporter
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"sort"
+	"sync"
+
+	"go.opencensus.io/trace"
+
+	commonpb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// Exporter is the type that converts OpenCensus Proto-Metrics into Prometheus metrics.
+type Exporter struct {
+	options   Options
+	handler   http.Handler
+	collector *collector
+	gatherer  prometheus.Gatherer
+}
+
+// Options customizes a created Prometheus Exporter.
+type Options struct {
+	Namespace   string
+	OnError     func(err error)
+	ConstLabels prometheus.Labels // ConstLabels will be set as labels on all views.
+	Registry    *prometheus.Registry
+}
+
+// New is the constructor to make an Exporter with the defined Options.
+func New(o Options) (*Exporter, error) {
+	if o.Registry == nil {
+		o.Registry = prometheus.NewRegistry()
+	}
+	collector := newCollector(o, o.Registry)
```

Comment:
Yes, one can choose to create a registry at startup time instead of using the default one. This was a requirement from opencensus-go/exporter/prometheus too.

Reply:
I'll actually just do
```go
registry := o.Registry
if registry == nil {
    registry = prometheus.NewRegistry()
}
```
to keep the clarity of the old.

Reply:
Okay now I recall why this was a problem. Internally Prometheus expects both the Options to have a non-nil Registry and then the collector to have a non-nil Registry. Otherwise we get this crash
```shell
=== RUN   TestPrometheusExporter_endToEnd
2019/03/12 17:51:19 http: panic serving 127.0.0.1:37978: runtime error: invalid memory address or nil pointer dereference
goroutine 50 [running]:
net/http.(*conn).serve.func1(0xc0000ca820)
	/home/travis/.gimme/versions/go1.11.5.linux.amd64/src/net/http/server.go:1746 +0x17c
panic(0xbd2340, 0x118fda0)
	/home/travis/.gimme/versions/go1.11.5.linux.amd64/src/runtime/panic.go:513 +0x1b9
github.com/prometheus/client_golang/prometheus.(*Registry).Gather(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/travis/gopath/pkg/mod/github.com/prometheus/client_golang@v0.9.1/prometheus/registry.go:407 +0x187
github.com/prometheus/client_golang/prometheus/promhttp.HandlerFor.func1(0xd2bfa0, 0xc0002d2000, 0xc0001e2700)
	/home/travis/gopath/pkg/mod/github.com/prometheus/client_golang@v0.9.1/prometheus/promhttp/http.go:105 +0x72
net/http.HandlerFunc.ServeHTTP(0xc00045f1d0, 0xd2bfa0, 0xc0002d2000, 0xc0001e2700)
	/home/travis/.gimme/versions/go1.11.5.linux.amd64/src/net/http/server.go:1964 +0x52
github.com/census-instrumentation/opencensus-service/exporter/prometheusexporter.(*Exporter).ServeHTTP(0xc00045f220, 0xd2bfa0, 0xc0002d2000, 0xc0001e2700)
	/home/travis/gopath/src/github.com/census-instrumentation/opencensus-service/exporter/prometheusexporter/exporter.go:158 +0x90
net/http.(*ServeMux).ServeHTTP(0xc000246b10, 0xd2bfa0, 0xc0002d2000, 0xc0001e2700)
	/home/travis/.gimme/versions/go1.11.5.linux.amd64/src/net/http/server.go:2361 +0x192
net/http.serverHandler.ServeHTTP(0xc0001ceea0, 0xd2bfa0, 0xc0002d2000, 0xc0001e2700)
	/home/travis/.gimme/versions/go1.11.5.linux.amd64/src/net/http/server.go:2741 +0xc5
net/http.(*conn).serve(0xc0000ca820, 0xd2cbe0, 0xc000469700)
	/home/travis/.gimme/versions/go1.11.5.linux.amd64/src/net/http/server.go:1847 +0x80b
created by net/http.(*Server).Serve
	/home/travis/.gimme/versions/go1.11.5.linux.amd64/src/net/http/server.go:2851 +0x4c6
--- FAIL: TestPrometheusExporter_endToEnd (0.00s)
    prometheus_test.go:163: Failed to perform a scrape: Get http://localhost:7777/metrics: EOF
```
so I'll revert this.